### PR TITLE
[C3PO-13] Add Code controllers implementations

### PR DESCRIPTION
**Pull request by saantiaguilera**
- Add GET implementation
- Add POST implementation
- Add PUT implementation
- Add golden file retriever
- Add CodeService abstraction (in a future this will be a DB + OS or
    something like that)
- Add unit tests of each file

### CODE REVIEW

File: src/api/controller/code/auth_middleware.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/arpb2/C-3PO/src/api/engine"
+	"github.com/gin-gonic/gin"
+	"net/http"
+	"strconv"
+)
+
+var AuthenticationMiddleware gin.HandlerFunc = handleAuthentication
+
+func handleAuthentication(ctx *gin.Context) {
+	authToken := ctx.GetHeader("Authentication")
```
Comment (mastanca) [inline src/api/controller/code/auth_middleware.go:14]: noted

File: src/api/controller/code/auth_middleware.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	if authToken == "" {
+		ctx.JSON(http.StatusUnauthorized, gin.H{
+			"error": "no 'Authentication' header provided",
```
Comment (mastanca) [inline src/api/controller/code/auth_middleware.go:18]: noted

File: src/api/golden/golden.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+)
+
+var update = flag.Bool("test.update", false, "update golden file")
+
+// Get returns the golden file content. If the `test.update` is specified, it updates the
+// file with the current output and returns it.
+func Get(t *testing.T, actual []byte, filename string) []byte {
```
Comment (mastanca) [inline src/api/golden/golden.go:14]: Para que es este file?

File: src/api/controller/code/auth_middleware.go (hunk 1 above)
Comment (saantiaguilera) [reply src/api/controller/code/auth_middleware.go:14]: Fixed on https://github.com/arpb2/C-3PO/commit/dc76453b9afd4ee19fcd4c737814632d29a6a0b9

File: src/api/controller/code/auth_middleware.go (hunk 2 above)
Comment (saantiaguilera) [reply src/api/controller/code/auth_middleware.go:18]: Fixed on https://github.com/arpb2/C-3PO/commit/dc76453b9afd4ee19fcd4c737814632d29a6a0b9

File: src/api/golden/golden.go (hunk 3 above)
Comment (saantiaguilera) [reply src/api/golden/golden.go:14]: Para poder hacer testeos con golden files (https://godoc.org/gotest.tools/golden). Golang ya te provee la API pero hice un wrapper para poder actualizar via flags / obtener sin tener que ensuciar todos los tests.

Googleatelo si no lo conocias (mas alla del framework de go) :)

**Pull request merged**